adapter/slack/classic_adapter.go

```diff
@@ -0,0 +1,445 @@
+
+			case *slack.InvalidAuthEvent:
+				e.Debug("Slack event: invalid auth")
+
+				events <- s.OnInvalidAuth(ev, info)
+
+				break eventLoop
+
+			case *slack.LatencyReport:
+				le := e.WithField("latency", ev.Value)
+				millis := ev.Value.Milliseconds()
+
+				switch {
+				case millis >= 1000 && millis < 1500:
+					le.Debug("Slack event: high latency detected")
+				case millis >= 1500 && millis < 2000:
+					le.Info("Slack event: high latency detected")
+				case millis >= 2000:
+					le.Warn("Slack event: high latency detected")
+				}
+
+			case *slack.MessageEvent:
+				providerEvent := s.OnMessage(ev, info)
+				if providerEvent != nil && providerEvent.EventType != "" {
+					events <- providerEvent
+				}
+
+			case *slack.RTMError:
+				e.WithError(ev).
+					WithField("code", ev.Code).
+					WithField("msg", ev.Msg).
+					Debug("Slack event: RTM error")
+				telemetry.Errors().WithError(ev).Commit(ctx)
+
+				events <- s.OnRTMError(ev, info)
+
+			case *slack.AckErrorEvent:
+				e.WithError(ev.ErrorObj).
+					WithField("replyto", ev.ReplyTo).
+					Error("Slack event: ACK event error")
+				telemetry.Errors().WithError(ev.ErrorObj).Commit(ctx)
+
+			case *slack.ConnectingEvent:
+				e.WithField("attempt", ev.Attempt).
+					WithField("connection.count", ev.ConnectionCount).
+					Debug("Slack event: connecting")
+
+			case *slack.IncomingEventError:
+				e.WithError(ev).
+					Debug("Slack event: error receiving incoming event")
+				telemetry.Errors().WithError(ev).Commit(ctx)
+
+			case *slack.OutgoingErrorEvent:
+				e.WithError(ev.ErrorObj).
+					WithField("message.channel", ev.Message.Channel).
+					WithField("message.id", ev.Message.ID).
+					WithField("message.text", ev.Message.Text).
+					WithField("message.type", ev.Message.Type).
+					Error("Slack event: outgoing message error")
+				telemetry.Errors().WithError(ev.ErrorObj).Commit(ctx)
+
+			case *slack.RateLimitedError:
+				e.WithError(ev).
+					WithField("retryafter", ev.RetryAfter).
+					Error("Slack event: API rate limited error")
+				telemetry.Errors().WithError(ev).Commit(ctx)
+
+			case *slack.UnmarshallingErrorEvent:
+				e.WithError(ev.ErrorObj).
+					Error("Slack event: failed to deconstruct Slack response")
+				telemetry.Errors().WithError(ev.ErrorObj).Commit(ctx)
+
+			case *slack.UserTypingEvent:
+				e.WithField("user", ev.User).
+					WithField("channel", ev.Channel).
+					Trace("Slack event: user is typing")
+
+			case *slack.HelloEvent:
+				// Do nothing (for now).
+
+			default:
+				// Report and ignore other events..
+				e.WithField("message.data", msg.Data).
+					WithField("type", fmt.Sprintf("%T", ev)).
+					Debug("Slack event: unhandled event type")
+			}
+		}
+
+		close(events)
+	}()
+
+	return events
+}
+
+// OnChannelMessage is called when the Slack API emits an MessageEvent for a message in a channel.
+func (s *ClassicAdapter) OnChannelMessage(event *slack.MessageEvent, info *adapter.Info) *adapter.ProviderEvent {
```

Comment:
Note for the future: these `OnFOO` methods are exported in the classic adapter (though they really don't need to be), and unexported in the new adapter. We might want to make them consistent eventually.

Reply:
Makes sense, do we want a follow up issue for it?

Reply:
Probably a good idea.